src/client/shadows/dynamicshadowsrender.cpp

```diff
@@ -0,0 +1,616 @@
+#include "client/shadows/dynamicshadowsrender.h"
+
+#include <type_traits>
+#include <utility>
+#include "settings.h"
+#include "filesys.h"
+#include "porting.h"
+#include "client/shader.h"
+
+#include "client/clientmap.h"
+
+ShadowRenderer::ShadowRenderer(irr::IrrlichtDevice *irrlichtDevice, Client *client) :
+	_device(irrlichtDevice), 
+	_smgr(irrlichtDevice->getSceneManager()),
+	_driver(irrlichtDevice->getVideoDriver()), 
+	_client(client) {
+	_shadows_enabled = g_settings->getBool("enable_shaders");
+	_shadows_enabled &= g_settings->getBool("enable_dynamic_shadows");
+	if (!_shadows_enabled)
+		return;
+
+	_shadow_strength = g_settings->getFloat("shadow_strength");
+
+	_shadow_map_max_distance = g_settings->getFloat("shadow_map_max_distance");
+
+	_shadow_map_texture_size = g_settings->getFloat("shadow_map_texture_size");
+
+	_shadow_map_texture_32bit = g_settings->getBool("shadow_map_texture_32bit");
+	_shadow_map_colored = g_settings->getBool("shadow_map_color");
+	_shadow_samples = g_settings->getS32("shadow_filters");
+	_shadow_psm = g_settings->getBool("shadow_psm");
+
+
+}
+
+ShadowRenderer::~ShadowRenderer() {
+	if (_shadow_depth_cb)
+		delete _shadow_depth_cb;
+	if (_shadow_mix_cb)
+		delete _shadow_mix_cb;
+	ShadowNodeArray.clear();
+	_light_list.clear();
+	// we don't have to delete the textures in renderTargets
+	/*
+	if (shadowMapTextureDynamicObjects) {
+		delete shadowMapTextureDynamicObjects;
+		shadowMapTextureDynamicObjects = nullptr;
+	}
+
+	if (shadowMapTextureFinal) {
+		delete shadowMapTextureFinal;
+		shadowMapTextureFinal = nullptr;
+	}
+	if (shadowMapTextureColors) {
+		delete shadowMapTextureColors;
+		shadowMapTextureColors = nullptr;
+	}
+	if (shadowMapClientMap) {
+		delete shadowMapClientMap;
+		shadowMapClientMap = nullptr;
+	}
+	*/
+}
+
+void ShadowRenderer::initialize() {
+	bool tempTexFlagMipMaps = _driver->getTextureCreationFlag(irr::video::ETCF_CREATE_MIP_MAPS);
+	bool tempTexFlag32 = _driver->getTextureCreationFlag(irr::video::ETCF_ALWAYS_32_BIT);
+	_driver->setTextureCreationFlag(irr::video::ETCF_CREATE_MIP_MAPS, tempTexFlagMipMaps);
+	_driver->setTextureCreationFlag(irr::video::ETCF_ALWAYS_32_BIT, tempTexFlag32);
+
+	irr::video::IGPUProgrammingServices *gpu = _driver->getGPUProgrammingServices();
+
+	// we need glsl
+	if (_shadows_enabled && gpu &&
+			(_driver->getDriverType() == irr::video::EDT_OPENGL &&
```

Comment:
are there concrete reasons why this couldn't work on GLESv2 or was that just not tested?

Reply:
Not tested. I'll figure out how can I test it.

Reply:
You don't need to take testing upon yourself, just removing the condition so it _can_ be tested would be fine.

Reply:
Ok. Done.